core/camel-support/src/main/java/org/apache/camel/support/EnumArray.java

```diff
@@ -0,0 +1,174 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.camel.support;
+
+import java.util.Arrays;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.Objects;
+import java.util.function.Function;
+
+/**
+ * This class provides an array-based structure used to store payloads associated with enums. This is
+ * used in the hot path of the core code to allow access to those payloads with constant time and low
+ * memory overhead.
+ *
+ * This data-structure is meant for internal usage of the Camel Core and is not meant for users.
+ * @param <T>
+ */
+public final class EnumArray<T extends Enum<?>> {
```

Comment:
```suggestion
public final class EnumArray<T extends Enum<T>> {
```

Reply:
Why not use `EnumMap` instead?

Reply:
@essobedo 🤔 I like your idea. Let me give it a try and see if we can skip the ssc issue with it and it may be a better implementation.